src/main/scala/dev/kamu/cli/utility/DependencyGraph.scala

```diff
@@ -0,0 +1,42 @@
+package dev.kamu.cli.utility
+
+import scala.collection.mutable
+
+class DependencyGraph(vertex_count: Int, deps: List[(Int, Int)]) {
+  // Dependency graph is a directed possibly disjoint graph without loops.
+  // TODO: Object creation should fail if the graph contains loops and report the loop as an error
+
+  // deps - directed edges of the graph in the form where
+  // (a, b) - vertex a depends on vertex b.
+
+  // add edges to the graph
+  private def addDep(v: Array[List[Int]], d: (Int, Int)) = v.updated(d._1, d._2 :: v.apply(d._1))
+  private val vertices: Array[List[Int]] = deps.foldLeft(Array.fill(vertex_count)(List.empty[Int]))(addDep)
+
+  def dependsOn: Int => List[Int] = vertices.apply
+
+  def resolve(goal: List[Int] /*, resolved: Array[Boolean]*/): Seq[Int] = {
+    val resolved: Array[Boolean] = Array.fill(vertex_count)(false)
+
+    val plan: mutable.Queue[Int] = new mutable.Queue()
+    val toSchedule: mutable.Stack[Int] = new mutable.Stack()
+    for (i <- goal) {
+      toSchedule.push(i)
+      while (toSchedule.nonEmpty) {
+        val toResolve: List[Int] = dependsOn(toSchedule.head).filterNot(resolved.apply)
+        if (toResolve.isEmpty) {
+          val x = toSchedule.pop()
+          if (!resolved(x)) {
+            plan.enqueue(x)
+            resolved(x) = true
+          }
+        }
+        else {
+          toSchedule.pushAll(toResolve)
+        }
+      }
+    }
+    plan
+  }
```

Comment:
It is simple indeed, but neither efficient nor correct.

This is not a sorting problem, thus sort might produce incorrect result. Lets say `A` depends on`B` and `C` depends on nothing. Imagine the user requested to pull `A,C,B`, none of the neighbors are connected, so the sorting algorithm might come to a conclusion the sequence `A,C,B` is already sorted. It just looks like `A=B=C`. It is not guaranteed to fail, since it depends on the exact implementation of the algorithm, yet it is possible to come up with an example for any predefined efficient sort algorithm. Only inefficient(at least quadratic in time) sorts, like bubble sort, would not suffer from this problem.

Recursive pull with dependency graph resolution is linear in time and space(compared to `n log^2 n` here). Not sure if linear time is possible with non-recursive reordering pull as in recursive one, but `n log n` is definitely possible.

Reply:
Yep, you're right it doesn't handle the disjoint case :(

Now that you simplified your algorithm a bit I can see what it's doing. You're starting at each input node and doing a traversal of its sub-graph while adding nodes you haven't seen before to the plan in depth-first order. This approach makes sense, but I still think we can simplify the implementation.

* We already have the `dependsOn` relationship captured in the `Dataset` itself so transforming the entire graph into a list of `(child, parent)` edges (as you currently do in `PullCommand`) only to later reconstruct it into `child -> list of parents` form seems redundant and this step can be eliminated.

* You are using a mutable stack to implement depth-first traversal. I think it's a lot more natural to implement it with recursion.

* I think you can avoid re-mapping datasetIDs into Ints and using indexes everywhere. The only two inputs this algorithm needs are: a sequence of DatasetIDs to resolve / order, and a `dependsOn` function that given a DatasetID returns list of its dependencies. And for keeping track of resolved nodes you could use a `Set`. You can still make the algorithm work on any graph-like objects if you want using generics. This approach also will lead you to a solution that loads the definitions of the datasets only for those datasets that matter.